src/mapper.rs

```diff
@@ -189,44 +192,250 @@ impl<'s> ProguardMapper<'s> {
     pub fn remap_frame(&'s self, frame: &StackFrame<'s>) -> RemappedFrameIter<'s> {
-        if let Some(class) = self.classes.get(frame.class.as_ref()) {
-            if let Some(members) = class.members.get(frame.method.as_ref()) {
+        if let Some(class) = self.classes.get(frame.class) {
+            if let Some(members) = class.members.get(frame.method) {
                 let mut frame = frame.clone();
-                frame.class = class.original.into();
+                frame.class = class.original;
                 return RemappedFrameIter::members(frame, members.iter());
             }
         }
         RemappedFrameIter::empty()
     }
 
+    /// Remaps a throwable which is the first line of a full stacktrace.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use proguard::{ProguardMapper,Throwable};
+    ///
+    /// let mapping = "com.example.Mapper -> a.b:";
+    /// let mapper = ProguardMapper::from(mapping);
+    ///
+    /// let throwable = Throwable::try_parse(b"a.b: Crash").unwrap();
+    /// let mapped = mapper.remap_throwable(&throwable);
+    ///
+    /// assert_eq!(Some(Throwable::with_message("com.example.Mapper", "Crash")), mapped);
+    /// ```
+    pub fn remap_throwable<'a>(&'a self, throwable: &Throwable<'a>) -> Option<Throwable<'a>> {
+        if let Some(class) = self.remap_class(throwable.class) {
+            Some(Throwable {
+                class,
+                message: throwable.message,
+            })
+        } else {
+            None
+        }
+    }
+
     /// Remaps a complete Java StackTrace.
-    pub fn remap_stacktrace(&'s self, input: &'s str) -> Result<String, std::fmt::Error> {
-        let mut stacktrace = String::new();
-        for line in input.lines() {
-            match StackFrame::try_parse(line.as_ref()) {
-                None => writeln!(&mut stacktrace, "{}", line)?,
-                Some(frame) => {
-                    let mut remapped = self.remap_frame(&frame).peekable();
-                    if remapped.peek().is_none() {
-                        writeln!(&mut stacktrace, "{}", line)?;
-                        continue;
-                    }
-                    for line in remapped {
-                        writeln!(
-                            &mut stacktrace,
-                            "    at {}.{}({}:{})",
-                            line.class,
-                            line.method,
-                            line.file.as_deref().unwrap_or("<unknown>"),
-                            line.line
-                        )?;
+    pub fn remap_stacktrace<'a>(&'a self, trace: &StackTrace<'a>) -> StackTrace<'a> {
+        let exception = trace
+            .exception
+            .as_ref()
+            .and_then(|t| self.remap_throwable(t));
+
+        let frames =
+            trace
+                .frames
+                .iter()
+                .fold(Vec::with_capacity(trace.frames.len()), |mut frames, f| {
```

Comment:
IMO a normal for loop is a bit more readable than a `fold`, but its fine either way.